memcached/init.lua

```diff
@@ -390,6 +405,17 @@ local function memcached_get(name)
     return memcached_services[name]
 end
 
+local function memcached_slab_info()
+    local slab_arena_info = memcached_internal.memcached_slab_arena_info()
+    local slab_info = {}
+    for _, v in pairs(slab_arena_info_table) do
+        slab_info[v] = tonumber(slab_arena_info[0][v])
```

Comment:
It is better to use `luaL_pushuint64()` for 64 bit numbers to don't loss precision on large values. The function is in the module API.

```lua
tarantool> tonumber(1152921504606846977ULL)
---
- 1152921504606846976
...
```

Reply:
With patch below 64bit numbers are used.

```diff
diff --git a/memcached/init.lua b/memcached/init.lua
index ce42edb..2b8626b 100644
--- a/memcached/init.lua
+++ b/memcached/init.lua
@@ -409,7 +409,7 @@ local function memcached_slab_info()
     local slab_arena_info = memcached_internal.memcached_slab_arena_info()
     local slab_info = {}
     for _, v in pairs(slab_arena_info_table) do
-        slab_info[v] = tonumber(slab_arena_info[0][v])
+        slab_info[v] = tonumber64(slab_arena_info[0][v])
     end
     slab_info['quota_used_ratio'] = string.format('%0.2f%%', slab_info['quota_used_ratio'])
 
diff --git a/test/common/module_api.test.lua b/test/common/module_api.test.lua
index 5676680..3bb172a 100755
--- a/test/common/module_api.test.lua
+++ b/test/common/module_api.test.lua
@@ -57,11 +57,11 @@ test:is(#memcached.server, 0, 'memcached.server is empty')
 -- memcached.slab(): memcached module is initialized, no instances
 
 test:istable(memcached.slab, 'memcached.slab (no instances) returns a table')
-test:isfunction(memcached.slab.info, 'memcached.slab.info (no instances) returns a function')
+test:isfunction(memcached.slab.info, 'memcached.slab.info() (no instances) returns a function')
 local slab_info = memcached.slab.info()
-test:isnumber(slab_info.quota_size, 'memcached.slab.info().quota_size (no instances) returns a number')
+test:is(type(slab_info.quota_size), 'cdata', 'memcached.slab.info().quota_size (no instances) returns a cdata')
 test:is(slab_info.quota_size, 4398046510080, 'memcached.slab.info().quota_size (no instances) is correct')
-test:isnumber(slab_info.quota_used, 'memcached.slab.info().quota_used (no instances) returns a number')
+test:is(type(slab_info.quota_used), 'cdata', 'memcached.slab.info().quota_used (no instances) returns a cdata')
 test:is(slab_info.quota_used, 0, 'memcached.slab.info().quota_used (no instances) is correct')
 test:isstring(slab_info.quota_used_ratio, 'memcached.slab.info().quota_used_ratio (no instances) returns a string')
 
@@ -126,17 +126,17 @@ mc_1:start()
 mc_2:start()
 
 test:istable(memcached.slab, 'memcached.slab returns a table')
-test:isfunction(memcached.slab.info, 'memcached.slab.info returns a function')
+test:isfunction(memcached.slab.info, 'memcached.slab.info() returns a function')
 local slab_info = memcached.slab.info()
 test:istable(slab_info, 'memcached.slab.info() returns a table')
-test:isnumber(slab_info.quota_size, 'memcached.slab.info().quota_size returns a number')
+test:is(type(slab_info.quota_size), 'cdata', 'memcached.slab.info().quota_size returns a cdata')
 test:is(slab_info.quota_size, 4398046510080, 'memcached.slab.info().quota_size is correct')
-test:isnumber(slab_info.quota_used, 'memcached.slab.info().quota_used returns a number')
+test:is(type(slab_info.quota_used), 'cdata', 'memcached.slab.info().quota_used returns a cdata')
 test:is(slab_info.quota_used, 4194304, 'memcached.slab.info().quota_used is correct')
 test:isstring(slab_info.quota_used_ratio, 'memcached.slab.info().quota_used_ratio returns a string')
-local ratio = 100 * slab_info.quota_used / slab_info.quota_size
-local actual_ratio = string.match(slab_info.quota_used_ratio, '^(%d.%d+)%%')
-local eps_is_correct = abs(actual_ratio - ratio) < 1
+local expected_ratio = tonumber(100 * slab_info.quota_used / slab_info.quota_size)
+local actual_ratio = tonumber(slab_info.quota_used_ratio:match('^(%d.%d+)%%'))
+local eps_is_correct = abs(actual_ratio - expected_ratio) < 1
 test:is(eps_is_correct, true, 'memcached.slab.info().quota_used_ratio is correct')
 
 mc_1:stop()
```

NOTE: `tonumber64()` for big numbers returns a value with type `cdata`, not a `number`. `tap` module has a separate check for `cdata` - `taptest:iscdata()`, but it doesn't work for me. Check failed with unclear error.

`ffi.lua`:

```lua
local memcached = require('memcached')
local tap = require('tap')
local memcached = require('memcached')
local test = tap.test('memcached module api')
local size = memcached.slab.info().quota_size
assert(type(size) == 'cdata')
test:iscdata(size, 'memcached.slab.info().quota_size (no instances) returns a cdata')
```

```sh
sergeyb@pony:~/sources/MRG/memcached$ tarantool ffi.lua 
creating arena with 4398046510080 bytes...
allocate slab cache with slab size 4194304
TAP version 13
LuajitError: builtin/tap.lua:207: declaration specifier expected near 'memcached'
fatal error, exiting the event loop
```

Reply:
Regarding `iscdata()`: you passed the message in place of ctype ('uint64' in the case).

Reply:
> Regarding iscdata(): you passed the message in place of ctype ('uint64' in the case).

Seems function prototype in [documentation](https://www.tarantool.io/en/doc/latest/reference/reference_lua/tap/#lua-function.taptest.iscdata) is incorrect: `taptest:iscdata(value, test-name)`.

Reply:
I got where I was wrong. I know that `tonumber64()` returns a value of type `'number'` for small numbers[^1], but I didn't know that `tonumber64()` returns cdata number for a small number passed as cdata number.

```lua
tarantool> type(tonumber64('0'))
---
- number
...

tarantool> type(tonumber64(0))
---
- number
...

tarantool> type(tonumber64(0ULL))
---
- cdata
...
```

That's why the test works: we just always return cdata numbers, at least on AMD64. `size_t` is 64 bit integer on AMD64 and the [FFI semantics](https://luajit.org/ext_ffi_semantics.html#convert_tolua) document states that LuaJIT converts those values to cdata numbers.

So `tonumber64()` is no-op. I would remove it. OTOH, it is harmless and I'll not insist.

Should we imitate `luaL_pushuint64()` behaviour, which is used in `box.slab.info()` and actually returns `'number'` for small values? I don't know a proper reason aside of 'most of tarantool functions behave this way'. So I'll not insist here too.

Since I have no strong agruments here, it is up to the author. I'll resolve the discussion.

[^1]: Here and below: less than or equal to 10^14-1.